pkg/virt-controller/watch/preset.go

```diff
@@ -0,0 +1,344 @@
+	recorder         record.EventRecorder
+	store            cache.Store
+}
+
+const initializerMarking = "presets.virtualmachines." + kubev1.GroupName + "/presets-applied"
+
+func NewVirtualMachinePresetController(vmPresetInformer cache.SharedIndexInformer, vmInitInformer cache.SharedIndexInformer, queue workqueue.RateLimitingInterface, vmInitCache cache.Store, clientset kubecli.KubevirtClient, recorder record.EventRecorder) *VirtualMachinePresetController {
+	vmi := VirtualMachinePresetController{
+		vmPresetInformer: vmPresetInformer,
+		vmInitInformer:   vmInitInformer,
+		clientset:        clientset,
+		queue:            queue,
+		recorder:         recorder,
+		store:            vmInitCache,
+	}
+	return &vmi
+}
+
+func (c *VirtualMachinePresetController) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting Virtual Machine Initializer.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmPresetInformer.HasSynced, c.vmInitInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping controller.")
+}
+
+func (c *VirtualMachinePresetController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *VirtualMachinePresetController) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VM %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VM %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VirtualMachinePresetController) execute(key string) error {
+
+	// Fetch the latest VM state from cache
+	obj, exists, err := c.store.GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	// If the VM isn't in the cache, it was just deleted, so shouldn't
+	// be initialized
+	if exists {
+		var vm *kubev1.VirtualMachine
+		vm = obj.(*kubev1.VirtualMachine)
+		// only process VM's that aren't initialized by this controller yet
+		if !isVirtualMachineInitialized(vm) {
+			return c.initializeVirtualMachine(vm)
+		}
+	}
+
+	return nil
+}
+
+func (c *VirtualMachinePresetController) initializeVirtualMachine(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	var err error
+
+	logger.Object(vm).Info("Initializing VirtualMachine")
+
+	allPresets, err := listPresets(c.vmPresetInformer, vm.GetNamespace())
+	if err != nil {
+		logger.Object(vm).Errorf("Listing VirtualMachinePresets failed: %v", err)
+		return err
+	}
+
+	matchingPresets := filterPresets(allPresets, vm, c.recorder)
+
+	if len(matchingPresets) != 0 {
+		applyPresets(vm, matchingPresets, c.recorder)
```

Comment:
You have to DeepCopy() the whole vm before you apply the presets. Otherwise you directly mutate the object inside the ache.